php/php-rr-sdk.md

```diff
@@ -0,0 +1,692 @@
+# PHP-SDK proposal
+The dynamic nature of PHP provides a wide range of possibilities how workflows can be defined or dynamically generated.
+A large variety of instruments, language performance and ease of use opens a lot of possibilities in a field of high-throughput, easy-to-develop
+distributed applications. Since PHP 5.5 we can use [cooperative multitasking](https://nikic.github.io/2012/12/22/Cooperative-multitasking-using-coroutines-in-PHP.html) 
+which can perfectly align with the workflow execution model.
+
+## Table of Contents
+- [Activities](#activities)
+    - [Registration](#registration)
+    - [Using Annotations](#using-annotations)
+    - [Logging](#logging)
+    - [Payloads](#payloads)
+    - [Process Isolation](#process-isolation)
+- [Workflows](#workflows)
+    - [Registration](#registration-1)
+    - [Using Annotations](#using-annotations-1)
+    - [Syntax and Atomic Blocks](#syntax-and-atomic-blocks)
+    - [Queries](#queries)
+    - [Signals](#signals)
+    - [Sessions](#sessions)
+    - [Deterministic Time](#deterministic-time)
+    - [Side Effects](#side-effects)
+    - [Error Handling](#error-handling)
+    - [Termination and Cancel Requests](#termination-and-cancel-requests)
+- [Examples](#examples)
+- [Service RPC](#service-rpc)
+- [Implementation](#implementation)
+    - [Features](#features)
+    - [Milestones](#milestones)
+- [Development](#development)
+
+## Activities
+Similar to other SDKs, the Activity implementation in PHP won't have any feasible limitations. Is it possible to invoke 
+any callable function, object method or anonymous function. The invocation arguments can be hydrated based on provided
+payload using [method reflection](https://www.php.net/manual/en/class.reflectionfunctionabstract.php) mechanism provided by php.    
+
+The invocation context and heartbeats could be addressed `ContextInterface` object. The error handling done using classic
+PHP exceptions.
+
+```php
+function processFile(string $url): string
+{
+    $result = doSomeWork($url);
+
+    return $result;
+}
+```
+
+Heartbeat example:
+
+```php
+use Temporal\Workflow\ContextInterface;
+use Temporal\Workflow\Activity;
+
+// parameter order does not matter
```

Comment:
Is it possible to use some sort of global/thread local variable to avoid passing context explicitly or it is PHP pattern [like in Go](https://blog.golang.org/context)?

Reply:
Yes and no. Practically we are going to run all workflows without one thread, so some _hidden_ context switching will be required. However, we can pass the context as part of the workflow constructor and mock it using easier abstraction.

To be honest, context is not really needed here since each of the workflows will be wrapped with a `yield` handler at the top (which manages the context).

So, technically, we can replace context with something like that:

```php
// was
$ctx->now();

// can be 
$now = yield new Now();

// or 
$now = yield $this->now();
```

Reply:
However, I think passing context is easier for 2 reasons:

1) It's more explicit
2) We can handle context passing in the constructor (since it's called only once) and hide the actual context from user code with base class implementation.